[Windows|macOS|Linux|BSD] - Can FTBFS on Clang due to C++17

> #### Checklist
> 
> <!-- You don't have to fill these out, but it would help us alot with managing open bug reports if you make sure to check these things first! -->
> 
> - [x] I am reporting exactly 1 bug with this issue.
> - [x] This bug hasn't already been reported.
> - [x] This bug hasn't already been fixed in the latest development build.

---

## Bug Description

Seeing this in Nixpkgs on the Qt6 macOS builds.

```
clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1z -pthread -Wall -Wextra -Wall -Wextra -Werror -pedantic -pedantic-errors -fPIC -DQT_DEPRECATED_WARNINGS -D__UNIX_JACK__ -DJACK_HAS_PORT_RENAME -D__MACOSX_CORE__ -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CORE5COMPAT_LIB -DQT_CORE_LIB -I. -Iinstrument -Imodule -I../submodules/emu2149/src -I../submodules/RtAudio/src -I/nix/store/6iqp86xqgp9nnhdrxgnzbrdjj0frx6cw-rtmidi-5.0.0/include/rtmidi -I/nix/store/qkrjmjlabb70am2s9nr0d2fq98m7xjd3-qtbase-6.7.0/lib/QtWidgets.framework/Headers -I/nix/store/qkrjmjlabb70am2s9nr0d2fq98m7xjd3-qtbase-6.7.0/lib/QtGui.framework/Headers -I/nix/store/n5y9vx5jsilvdxgwaraffln9rg48g0ix-qt5compat-6.7.0/lib/QtCore5Compat.framework/Headers -I/nix/store/qkrjmjlabb70am2s9nr0d2fq98m7xjd3-qtbase-6.7.0/lib/QtCore.framework/Headers -I. -I/System/Library/Frameworks/OpenGL.framework/Headers -I/System/Library/Frameworks/AGL.framework/Headers/ -I. -I/System/Library/Frameworks/OpenGL.framework/Headers -I/System/Library/Frameworks/AGL.framework/Headers/ -I/nix/store/qkrjmjlabb70am2s9nr0d2fq98m7xjd3-qtbase-6.7.0/mkspecs/macx-clang -F/nix/store/qkrjmjlabb70am2s9nr0d2fq98m7xjd3-qtbase-6.7.0/lib -F/nix/store/n5y9vx5jsilvdxgwaraffln9rg48g0ix-qt5compat-6.7.0/lib -o instruments_manager.o instrument/instruments_manager.cpp
In file included from instrument/instruments_manager.cpp:35:
./utils.hpp:96:26: error: 'result_of<(lambda at instrument/instruments_manager.cpp:733:2) &&(std::shared_ptr<AbstractInstrument>)>' is deprecated [-Werror,-Wdeprecated-declarations]
        using T = typename std::result_of<UnaryOperation&&(typename InputContainer::value_type)>::type;
                                ^
instrument/instruments_manager.cpp:732:16: note: in instantiation of function template specialization 'utils::transformIf<std::vector, std::array<std::shared_ptr<AbstractInstrument>, 128>, (lambda at instrument/instruments_manager.cpp:732:36), (lambda at instrument/instruments_manager.cpp:733:2)>' requested here
        return utils::transformIf(insts_, [](InstRef inst) { return inst; },
                      ^
/nix/store/xg2m3mbydbkvbg4zb86zvgg4c35hj1mi-libcxx-16.0.6-dev/include/c++/v1/__type_traits/result_of.h:24:34: note: 'result_of<(lambda at instrument/instruments_manager.cpp:733:2) &&(std::shared_ptr<AbstractInstrument>)>' has been explicitly marked deprecated here
template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                 ^
/nix/store/xg2m3mbydbkvbg4zb86zvgg4c35hj1mi-libcxx-16.0.6-dev/include/c++/v1/__config:808:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/nix/store/xg2m3mbydbkvbg4zb86zvgg4c35hj1mi-libcxx-16.0.6-dev/include/c++/v1/__config:781:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
1 error generated.
make[1]: *** [Makefile:5627: instruments_manager.o] Error 1
```

We request C++14, but Qt6 demands C++17 or greater. And Clang seems very strict about this deprecation, which triggers `-Werror` to abort the build.

There are already comments in the code that acknowledge this deprecation:

https://github.com/BambooTracker/BambooTracker/blob/e91eb3f207e74dd36cdd3e9d1c913954a179e6d9/BambooTracker/utils.hpp#L95-L96

Example log: https://hydra.nixos.org/build/256589183/nixlog/1

## How to reproduce

<!-- If applicable, give a step-by-step guide on how to reproduce your bug. If you encounter problems with any files in particular, please attach them. -->

1. `nix-build -A nixpkgs.bambootracker-qt6` on macOS with Nix installed

## System Information

<!-- Please fill out these details about your system, to help us debug your problem. -->

- **Operating System**: macOS
- **BambooTracker Version**: 0.6.3
- **Build Type**: Nix Package Manager

Just noting that this can happen when building with Qt5 as well now.

```
In file included from instrument/instruments_manager.cpp:35:
./utils.hpp:96:26: error: 'result_of<(lambda at instrument/instruments_manager.cpp:733:2) &&(std::shared_ptr<AbstractInstrument>)>' is deprecated [-Werror,-Wdeprecated-declarations]
        using T = typename std::result_of<UnaryOperation&&(typename InputContainer::value_type)>::type;
                                ^
instrument/instruments_manager.cpp:732:16: note: in instantiation of function template specialization 'utils::transformIf<std::vector, std::array<std::shared_ptr<AbstractInstrument>, 128>, (lambda at instrument/instruments_manager.cpp:732:36), (lambda at instrument/instruments_manager.cpp:733:2)>' requested here
        return utils::transformIf(insts_, [](InstRef inst) { return inst; },
                      ^
/nix/store/14bkhpbbl0rqzlb54y9v0j2sfx0pkc32-libcxx-16.0.6-dev/include/c++/v1/__type_traits/result_of.h:24:34: note: 'result_of<(lambda at instrument/instruments_manager.cpp:733:2) &&(std::shared_ptr<AbstractInstrument>)>' has been explicitly marked deprecated here
template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                 ^
/nix/store/14bkhpbbl0rqzlb54y9v0j2sfx0pkc32-libcxx-16.0.6-dev/include/c++/v1/__config:808:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/nix/store/14bkhpbbl0rqzlb54y9v0j2sfx0pkc32-libcxx-16.0.6-dev/include/c++/v1/__config:781:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
```

(`CONFIG+=no_warnings_are_errors` or `-DWARNINGS_ARE_ERRORS:BOOL=OFF` to work around it, ofc)

--- a/BambooTracker/utils.hpp
+++ b/BambooTracker/utils.hpp
@@ -72,8 +72,7 @@ template <template <typename ...> class OutputContainer = std::vector,
 		  class InputContainer, class UnaryOperation>
 inline auto transform(InputContainer& input, UnaryOperation&& op)
 {
-	// std::result_of is deprecated in C++17 and obsoluted in C++20, use std::invoke_result after C++14
-	using T = typename std::result_of<UnaryOperation&&(typename InputContainer::value_type)>::type;
+	using T = typename std::invoke_result<UnaryOperation&&,typename InputContainer::value_type>::type;
 	OutputContainer<T> output;
 	std::transform(input.begin(), input.end(), std::back_inserter(output), op);
 	return output;
@@ -92,8 +91,7 @@ template <template <typename ...> class OutputContainer = std::vector,
 		  class InputContainer, class Predicate, class UnaryOperation>
 inline auto transformIf(const InputContainer& input, Predicate&& pred, UnaryOperation&& op)
 {
-	// std::result_of is deprecated in C++17 and obsoluted in C++20, use std::invoke_result after C++14
-	using T = typename std::result_of<UnaryOperation&&(typename InputContainer::value_type)>::type;
+	using T = typename std::invoke_result<UnaryOperation&&,typename InputContainer::value_type>::type;
 	OutputContainer<T> output;
 	transformIf(input.begin(), input.end(), std::back_inserter(output), pred, op);
 	return output;

